Add table of issue type info and relevant column name descriptions (#1100)
cleanlab/cleanlab: docs/source/cleanlab/datalab/guide/issue_type_description.rst

In case you didn't know: you can alternatively use `Cleanlab Studio <https://cleanlab.ai/blog/data-centric-ai/>`_ to detect the same data issues as this package, plus `many more types of issues <https://help.cleanlab.ai/guide/concepts/cleanlab_columns/>`_, all without having to do any Machine Learning (or even write any code).

.. include:: table.rst

> 

Reviewer: let's add the table, it's awesome!

Some metadata about label issues is stored in the `issues` attribute of the Datalab object.
Let's look at one way to access this information.

.. testcode::
    
    lab.get_issues("label").sort_values("label_score").head(5)

The output will look something like this:

.. testoutput::

        is_label_issue  label_score  given_label  predicted_label
    97            True     0.064045            0                2
    58           False     0.680894            2                2
    41           False     0.746043            0                0
    4            False     0.794894            2                2
    98           False     0.802911            1                1

``is_label_issue``
~~~~~~~~~~~~~~~~~~

A boolean column that flags examples with label issues. 
If `True`, the example is estimated to have a label issue.
If `False`, the example is estimated to not have a label issue.

``label_score``
~~~~~~~~~~~~~~~

A numeric column that gives the label quality score for each example.
The score lies between 0 and 1.
The lower the score, the less likely the given label is to be correct.

``given_label``
~~~~~~~~~~~~~~~

> A column of the actual labels as provided in the dataset.

Reviewer: ```suggestion
A column of the actual labels as provided in the original dataset.
```

Learn more about the methods used to detect outliers in our article: `Out-of-Distribution Detection via Embeddings or Predictions <https://cleanlab.ai/blog/outlier-detection/>`_

Some metadata about outlier issues is stored in the `issues` attribute of the Datalab object.
Let's look at one way to access this information.

.. testcode::

    lab.get_issues("outlier").sort_values("outlier_score").head(5)

The output will look something like this:

.. testoutput::

        is_outlier_issue  outlier_score
    98              True       0.011562
    62             False       0.019657
    22             False       0.035243
    1              False       0.040907
    42             False       0.056865

``is_outlier_issue``
~~~~~~~~~~~~~~~~~~~~

A boolean column, where `True` indicates that an example is identified as an outlier.

``outlier_score``
~~~~~~~~~~~~~~~~~

A numeric column with scores between 0 and 1. 
A smaller value for an example indicates that it is less common or typical in the dataset, suggesting that it is more likely to be an outlier.

> If most of the nearest-neighbors of an example are exact-duplicates, then the outlier score of the example is set to 1.0.

Reviewer: ```suggestion
```

        is_near_duplicate_issue  near_duplicate_score near_duplicate_sets distance_to_nearest_neighbor  
    36                     True              0.066009            [11, 80]                     0.003906    
    11                     True              0.066009                [36]                     0.003906    
    80                     True              0.093245                [36]                     0.005599    
    27                    False              0.156720                  []                     0.009751    
    72                    False              0.156720                  []                     0.009751    

``is_near_duplicate_issue``
~~~~~~~~~~~~~~~~~~~~~~~~~~~

A boolean column, where `True` indicates that an example is identified as either a near- or exact-duplicate of other examples in the dataset.

``near_duplicate_score``
~~~~~~~~~~~~~~~~~~~~~~~~

A numeric column with scores between 0 and 1. The lower the score, the more likely the example is to be a near-duplicate of another example in the dataset.

Exact duplicates are assigned a score of 0, while near-duplicates are assigned a score close to 0.

``near_duplicate_sets``
~~~~~~~~~~~~~~~~~~~~~~~

> A column of lists of integers, where each list contains the indices of examples that belong to the same set of near-duplicates (not including the example itself).
> Each set represents a group of examples that are extremely similar to each other, relative to the rest of the dataset.
> The examples in each set may be exactly duplicated or have very similar feature representations.

Reviewer: ```suggestion
A column of lists of integers. The i-th list contains the indices of examples that are considered near-duplicates of example i (not including example i).
```

``distance_to_nearest_neighbor``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> A numeric column that represents the distance between each example and its nearest neighbor in the dataset.
> The distance is calculated based on the provided `features` or `knn_graph`.
> A smaller distance indicates that the example is more similar to its nearest neighbor.
> Examples that are (near) duplicates have smaller distances to their nearest neighbors compared to other examples in the dataset.
> Exact duplicates ideally have a distance of 0 to their nearest neighbor. However, due to floating point precision, especially when using certain distance metrics like Euclidean distance, this might not always be the case.

Reviewer: ```suggestion
A numeric column that represents the distance between each example and its nearest neighbor in the dataset.
The distance is calculated based on the provided `features` or `knn_graph`, and is directly related to the `near_duplicate_score`.
A smaller distance indicates that the example is more similar to its nearest neighbor in the dataset.
```

> Class imbalance in a dataset can lead to subpar model performance for the under-represented class. Consider collecting more data from the under-represented class, or at least take special care while modeling via techniques like over/under-sampling, SMOTE, asymmetric class weighting, etc.

Reviewer: ```suggestion
Class imbalance in a dataset can lead to subpar model performance for the under-represented class. Consider collecting more data from the under-represented class, or at least take special care while modeling via techniques like over/under-sampling, SMOTE, asymmetric class weighting, etc.

This issue-type is more about the overall dataset vs. individual data points. If severe class imbalance is detected, Datalab will flag the individual data points from the minority class.
```

Some metadata about class imbalance issues is stored in the `issues` attribute of the Datalab object.
Let's look at one way to access this information.

.. testcode::

    lab.get_issues("class_imbalance").sort_values("class_imbalance_score").head(5)

The output will look something like this:

.. testoutput::

        is_class_imbalance_issue  class_imbalance_score  given_label
    27                     False                   0.28            2
    72                     False                   0.28            2
    75                     False                   0.28            2
    33                     False                   0.28            2
    68                     False                   0.28            2

``is_class_imbalance_issue``
~~~~~~~~~~~~~~~~~~~~~~~~~~~~

> 
> A boolean column, where `True` indicates that the dataset suffers class imbalance. All examples belonging to the rarest class are flagged with this issue.
> But the dataset is considered to have class imbalance only if the proportion of examples belonging to the rarest class must be much lower than the inverse of the number of classes in the dataset.

Reviewer: ```suggestion

A boolean column, where `True` indicates which data points belong to the minority class (rarest class) in a classification dataset that exhibits severe class imbalance.  If the dataset is not considered to have severe class imbalance (i.e. proportion of examples in the rarest class is not to small relative to the number of classes in the dataset), then all values will be `False`.

```

``class_imbalance_score``
~~~~~~~~~~~~~~~~~~~~~~~~~

> A numeric column with scores between 0 and 1.
> Any example belonging to the most under-represented class is assigned a score equal to the proportion of examples in the dataset belonging to that class.
> All other examples are assigned a score of 1.

Reviewer: ```suggestion
A numeric column with scores between 0 and 1.
Any example belonging to the most under-represented class is assigned a score equal to the proportion of examples in the dataset belonging to that class.
All other examples are assigned a score of 1.
All examples sharing the same label also share the same score.
```

``given_label``
~~~~~~~~~~~~~~~

> A column of the actual labels as provided in the dataset.

Reviewer: ```suggestion
A column of the actual labels as provided in the original dataset.
```

Presence of null examples in the dataset can lead to errors when training ML models. It can also
result in the model learning incorrect patterns due to the null values.

Some metadata about null issues is stored in the `issues` attribute of the Datalab object.
Let's look at one way to access this information.

.. testcode::

    lab.get_issues("null").sort_values("null_score").head(5)

The output will look something like this:

.. testoutput::

        is_null_issue  null_score
    0           False         1.0
    72          False         1.0
    71          False         1.0
    70          False         1.0
    69          False         1.0

``is_null_issue``
~~~~~~~~~~~~~~~~~

A boolean column, where `True` indicates that an example is identified as having null/missing values across all feature columns.

> Even if an example has a single non-null value, it is not flagged with this issue.

Reviewer: ```suggestion
Examples that just have a single non-null value across multiple feature columns are not flagged with this issue.
```

``null_score``
~~~~~~~~~~~~~~

> A numeric column with scores between 0 and 1. The score represents the proportion of non-null values in the example.
> A lower score indicates that the example has more null/missing values.
> A score of 0 indicates that all feature values in the example are null. These examples are flagged with the null issue.
> A score of 1 indicates that none of the feature values in the example are null.

Reviewer: ```suggestion
A numeric column with scores between 0 and 1. The score represents the proportion of non-null (i.e. non-missing) values in each example.
Lower scores indicate examples with more null/missing values.
```

The data valuation score is an approximate Data Shapley value, calculated based on the labels of the top k nearest neighbors of an example. The details of this KNN-Shapley value could be found in the papers: `Efficient Task-Specific Data Valuation for Nearest Neighbor Algorithms <https://arxiv.org/abs/1908.08619>`_ and `Scalability vs. Utility: Do We Have to Sacrifice One for the Other in Data Importance Quantification? <https://arxiv.org/abs/1911.07128>`_.

Some metadata about data valuation issues is stored in the `issues` attribute of the Datalab object.
Let's look at one way to access this information.

.. testcode::

    lab.get_issues("data_valuation").sort_values("data_valuation_score").head(5)

The output will look something like this:

.. testoutput::

        is_data_valuation_issue  data_valuation_score
    39                    False                   0.5
    32                    False                   0.5
    98                    False                   0.5
    6                     False                   0.5
    7                     False                   0.5

``is_data_valuation_issue``
~~~~~~~~~~~~~~~~~~~~~~~~~~~

> A boolean column, where `True` indicates that an example is contributes negatively to a model's training performance.

Reviewer: ```suggestion
A boolean column, where `True` indicates that an example does not appear to contribute positively to a model's training performance.
```

``data_valuation_score``
~~~~~~~~~~~~~~~~~~~~~~~~

> A numeric column with scores between 0 and 1. The score reflects how valuable an example is in terms of improving or maintaining the performance of the model during training.
> A score below 0.5 indicates that the example contributes negatively to the model's performance. Such examples are flagged as having a data valuation issue.
> A score above 0.5 indicates that the example positively influences the model's performance, contributing to better learning and predictive accuracy.

Reviewer: ```suggestion
A numeric column with scores between 0 and 1. The score reflects how valuable each individual example is in terms of improving the performance of the ML model trained on this dataset.
Examples with higher scores more positively influence the resulting model's predictive performance, contributing to better learning. One would expect the model to get worse if many such examples were removed from its training dataset.
```
